### PR TITLE
Embed Source Metadata in App

### LINKED ISSUE #1137: Provide project source metadata from local git information
### Description
As a buildpacks user, I'd like to have the source and version information of my project embedded as metadata in the final app image.

### Proposed solution

Based on the spec, we should be able to provide `<layers>/project-metadata.toml` in [this format](https://github.com/buildpacks/spec/blob/platform/0.6/platform.md#project-metadatatoml-toml) which would be translated into the [`io.buildpacks.project.metadata` label](https://github.com/buildpacks/spec/blob/platform/0.6/platform.md#iobuildpacksprojectmetadata-json) on the image.

This issue should be scoped to local git information. 

##### General case

```gherkin
Given I clone a project from git@github.com:buildpacks/pack.git
And `git rev-parse HEAD` returns "8348484564e6aa0699de0ee78c258e88240eb0b5"
And `git describe --tags --always` returns "v0.18.1-2-g83484845"
And `git branch --show-current` returns "main"
And `git remote -v | grep "fetch"` returns:
    "origin	git@github.com:buildpacks/pack.git (fetch)"
And I run `pack build my-app`
When I run the following command `docker inspect my-app | jq -r '.[0].Config.Labels["io.buildpacks.project.metadata"] | fromjson'`
Then I should see:
{
  "type": "git",
  "version": { "commit": "8348484564e6aa0699de0ee78c258e88240eb0b5", "describe": "v0.18.1-2-g83484845"},
  "metadata": {
    "refs": ["main"],
    "url": "git@github.com:buildpacks/pack.git"
  }
}
```

##### At a tag(s)

```gherkin
Given I clone a project from git@github.com:buildpacks/pack.git
And `git rev-parse HEAD` returns "8348484564e6aa0699de0ee78c258e88240eb0b5"
And `git describe --tags --always` returns "v0.18.1"
And `git branch --show-current` returns "main"
And `git tag --points-at HEAD` returns:
  v0.18.1
  v0.18.1-rc2
And `git remote -v | grep "fetch"` returns:
    "origin	git@github.com:buildpacks/pack.git (fetch)"
And I run `pack build my-app`
When I run the following command `docker inspect my-app | jq -r '.[0].Config.Labels["io.buildpacks.project.metadata"] | fromjson'`
Then I should see:
{
  "type": "git",
  "version": { "commit": "8348484564e6aa0699de0ee78c258e88240eb0b5", "describe": "v0.18.1"},
  "metadata": {
    "refs": ["main", "v0.18.1", "v0.18.1-rc2"],
    "url": "git@github.com:buildpacks/pack.git"
  }
}
```

### Additional context

This issue is part of an epic of work. The related issues may be see by filtering on [epic/source-metadata](https://github.com/buildpacks/pack/labels/epic%2Fsource-metadata).

### PR DESCRIPTION
**Pull request by Haimantika**
## Summary
Provide project source metadata from local git information (#1137)

## Documentation

- Should this change be documented?
    - [ ] Yes, see #___
    - [ ] No

## Related

Resolves #1137

### CODE REVIEW

**Comment by dfreilich** [comment]
Hey @Haimantika , thank you so much for the work! The test you added isn't passing, both locally (for me) and on our CI. Do you think you'd be able to take another look at it and get it passing?

**Comment by Haimantika** [comment]
> Hey @Haimantika , thank you so much for the work! The test you added isn't passing, both locally (for me) and on our CI. Do you think you'd be able to take another look at it and get it passing?

Hi David, yes I acknowledge it. I am trying to solve the errors locally, will probably have an updated test soon.

**Comment by dfreilich** [comment]
Fantastic! Also, it looks like some of the changes from #1268 weren't pulled into this PR. Just wanted to make sure that they're synced up. I'll close that in favor of this.

File: build.go [hunk 1]
```diff
@@ -296,25 +298,46 @@ func (c *Client) Build(ctx context.Context, opts BuildOptions) error {
...
-	lifecycleOpts := build.LifecycleOptions{
-		AppPath:        appPath,
-		Image:          imageRef,
-		Builder:        ephemeralBuilder,
-		LifecycleImage: ephemeralBuilder.Name(),
-		RunImage:       runImageName,
-		ProjectMetadata: platform.ProjectMetadata{Source: &platform.ProjectSource{
+	r, error := git.PlainOpen(appPath)
+
+	if error != nil {
```
Comment (jromero) [inline build.go:314]: 1. This logic seems to be inverted. I believe you want to use type=project when error is "not found".
2. We'll want to add a check that `ProjectDescriptor` is not nil.

File: build.go [hunk 2]
```diff
@@ -296,25 +298,46 @@ func (c *Client) Build(ctx context.Context, opts BuildOptions) error {
...
+		MetadataObj = platform.ProjectSource{
 			Type:     "project",
 			Version:  map[string]interface{}{"declared": version},
-			Metadata: map[string]interface{}{"url": sourceURL},
-		}},
+			Metadata: map[string]interface{}{"url": sourceURL}}
+	} else {
+		hashCommit, _ := r.Head()
+		tagObj, _ := r.TagObject(hashCommit.Hash())
```
Comment (jromero) [inline build.go:321]: Ignoring errors is not acceptable in this scenario.

**Comment by jkutner** [comment]
I'm going to close this as it's become fairly out of date. Let us know if we can help move it forward.

**Pull request closed without merging**